test/nodetool/test_snapshot.py

```diff
@@ -99,3 +102,195 @@ def test_listsnapshots(nodetool, request):
         assert res == scylla_expected_output
     else:
         assert res == cassandra_expected_output
+
+
+def check_snapshot_out(res, tag, ktlist, skip_flush):
+    if type(tag) is int:
+        tag_pattern = str(tag)[:-5] + "[0-9]{5}"
```

Comment:
this introduces a flaky test. here is a real world example: tag = `1699251599822`. and after a while, we would have a snapshot whose tag is `1699251600039`, please note
`1699251600039-1699251599822` is just 217.

see https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4530/testReport/junit/test.nodetool/test_snapshot/Tests___Unit_Tests___test_snapshot_options_matrix_False_ktlist5_None_/ and https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/4530/testReport/junit/test.nodetool/test_snapshot/Tests___Unit_Tests___test_snapshot_options_matrix_False_ktlist5_None__2/

Reply:
```
AssertionError: assert None
 +  where None = <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00>('Requested creating snapshot(s) for [ks1,ks2] with snapshot name [1699251600039] and options {skipFlush=false}')
 +    where <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00> = re.compile('Requested creating snapshot\\(s\\) for \\[ks1, ?ks2\\] with snapshot name \\[16992515[0-9]{5}\\] and options \\{skipFlush=false\\}').fullmatch
..
nodetool = <function nodetool.<locals>.invoker at 0x7f6c8386a5c0>
tag = 1699251599822, ktlist = {'ks': ['ks1', 'ks2'], 'tbl': []}
skip_flush = False

    @pytest.mark.parametrize("tag", [None, "my_snapshot_tag"])
    @pytest.mark.parametrize("ktlist", [
        {},
        {"ks": ["ks.tbl1"]},
        {"ks": ["ks1.tbl1", "ks2.tbl2"]},
        {"ks": ["ks1"], "tbl": ["tbl1"]},
        {"ks": ["ks1"], "tbl": ["tbl1", "tbl2"]},
        {"ks": ["ks1", "ks2"], "tbl": []},
    ])
    @pytest.mark.parametrize("skip_flush", [False, True])
    def test_snapshot_options_matrix(nodetool, tag, ktlist, skip_flush):
        cmd = ["snapshot"]
        params = {}
    
        if tag is None:
            tag = int(time.time() * 1000)
            params["tag"] = approximate_value(value=tag, delta=99999).to_json()
        else:
            cmd += ["--tag", tag]
            params["tag"] = tag
    
        if skip_flush:
            cmd.append("--skip-flush")
    
        params["sf"] = str(skip_flush).lower()
    
        if ktlist:
            if "tbl" in ktlist:
                if len(ktlist["tbl"]) > 0:
                    keyspaces = ktlist["ks"]
                    cmd += ["--table", ",".join(ktlist["tbl"])]
                    cmd += keyspaces
                    params["kn"] = keyspaces[0]
                    params["cf"] = ",".join(ktlist["tbl"])
                else:
                    keyspaces = ktlist["ks"]
                    cmd += keyspaces
                    params["kn"] = ",".join(keyspaces)
            else:
                keyspaces = ktlist["ks"]
                cmd += ["-kt", ",".join(keyspaces)]
                if len(keyspaces) == 1:
                    ks, tbl = keyspaces[0].split(".")
                    params["kn"] = ks
                    params["cf"] = tbl
                elif len(keyspaces) > 1:
                    params["kn"] = ",".join(keyspaces)
        else:
            keyspaces = []
    
        res = nodetool(*cmd, expected_requests=[
            expected_request("POST", "/storage_service/snapshots", params=params)
        ])
    
>       check_snapshot_out(res, tag, keyspaces, skip_flush)

test/nodetool/test_snapshot.py:238: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

res = 'Requested creating snapshot(s) for [ks1,ks2] with snapshot name [1699251600039] and options {skipFlush=false}\nSnapshot directory: 1699251600039\n'
tag = 1699251599822, ktlist = ['ks1', 'ks2'], skip_flush = False

    def check_snapshot_out(res, tag, ktlist, skip_flush):
        if type(tag) is int:
            tag_pattern = str(tag)[:-5] + "[0-9]{5}"
        else:
            tag_pattern = tag
    
        if len(ktlist) == 0:
            keyspaces = "all keyspaces"
        else:
            keyspaces = ", ?".join(ktlist)
    
        main_pattern = re.compile("Requested creating snapshot\\(s\\)"
                                  f" for \\[{keyspaces}\\]"
                                  f" with snapshot name \\[{tag_pattern}\\]"
                                  f" and options \\{{skipFlush={str(skip_flush).lower()}\\}}")
    
        snapshot_dir_pattern = re.compile(f"Snapshot directory: {tag_pattern}")
    
        print(res)
    
        lines = res.split("\n")
        assert len(lines) == 3
>       assert main_pattern.fullmatch(lines[0])
E       AssertionError: assert None
E        +  where None = <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00>('Requested creating snapshot(s) for [ks1,ks2] with snapshot name [1699251600039] and options {skipFlush=false}')
E        +    where <built-in method fullmatch of re.Pattern object at 0x5644cc4d8e00> = re.compile('Requested creating snapshot\\(s\\) for \\[ks1, ?ks2\\] with snapshot name \\[16992515[0-9]{5}\\] and options \\{skipFlush=false\\}').fullmatch

test/nodetool/test_snapshot.py:129: AssertionError
```

Reply:
Wow, I knew I didn't understand what this trick was trying to do (I said: "I have no idea what this weird construct does, or actually what this function wants to do or what its "tag" parameter actually is. ) but I missed that it's a downright bug.
@denesb I'll dequeue the PR.

Reply:
probably we can use the homebrew `approximate_value` helper or pytest.approx for testing this? but i'd still like to understand what the integer represented by "tag" is for and why we use "5" here. probably because its a timestamp in microsecond?

Reply:
Fixed in v2. I now extract the tag and match it according to the type.